skfda/preprocessing/dim_reduction/projection/_fpca.py

```diff
@@ -1,33 +1,433 @@
+            else X.basis.n_basis
+        n_samples = X.n_samples
+
+        # check that the number of components is smaller than the sample size
+        if self.n_components > X.n_samples:
+            raise AttributeError("The sample size must be bigger than the "
+                                 "number of components")
+
+        # check that we do not exceed limits for n_components as it should
+        # be smaller than the number of attributes of the basis
+        if self.n_components > n_basis:
+            raise AttributeError("The number of components should be "
+                                 "smaller than the number of attributes of "
+                                 "target principal components' basis.")
+
+        # if centering is True then subtract the mean function to each function
+        # in FDataBasis
+        if self.centering:
+            meanfd = X.mean()
+            # consider moving these lines to FDataBasis as a centering function
+            # subtract from each row the mean coefficient matrix
+            X.coefficients -= meanfd.coefficients
+
+        # setup principal component basis if not given
+        if self.components_basis:
+            # First fix domain range if not already done
+            self.components_basis.domain_range = X.basis.domain_range
+            g_matrix = self.components_basis.gram_matrix()
+            # the matrix that are in charge of changing the computed principal
+            # components to target matrix is essentially the inner product
+            # of both basis.
+            j_matrix = X.basis.inner_product(self.components_basis)
+        else:
+            # if no other basis is specified we use the same basis as the passed
+            # FDataBasis Object
+            self.components_basis = X.basis.copy()
+            g_matrix = self.components_basis.gram_matrix()
+            j_matrix = g_matrix
+
+        # make g matrix symmetric, referring to Ramsay's implementation
+        g_matrix = (g_matrix + np.transpose(g_matrix)) / 2
+
+        # Apply regularization / penalty if applicable
+        if self.regularization_parameter > 0:
+            # obtain regularization matrix
+            regularization_matrix = self.components_basis.penalty(
+                self.regularization_derivative_degree,
+                self.regularization_coefficients)
+            # apply regularization
+            g_matrix = g_matrix + self.regularization_parameter \
+                * regularization_matrix
+
+        # obtain triangulation using cholesky
+        l_matrix = np.linalg.cholesky(g_matrix)
+
+        # we need L^{-1} for a multiplication, there are two possible ways:
+        # using solve to get the multiplication result directly or just invert
+        # the matrix. We choose solve because it is faster and more stable.
+        # The following matrix is needed: L^{-1}*J^T
+        l_inv_j_t = np.linalg.solve(l_matrix, np.transpose(j_matrix))
```

Comment:
Is it not `scipy.linalg.cho_solve` faster?

Reply:
I forgot to change this to solve_triangular, which is the fastest of the methods I have compared.